tests/test_actual_langservers.py

```diff
@@ -0,0 +1,393 @@
+METHOD_COMPLETION = "completion"
+RESPONSE_TYPES = {METHOD_COMPLETION: lsp.Completion}
+
+
+def find_method_marker(text, method):
+    """ searches for line: `<code> #<method>-<shift>`
+          - example: `sys.getdefaultencoding() #{METHOD_COMPLETION}-5`
+            position returned will be 5 chars before `#...`: `sys.getdefaultencodi | ng()`
+    """
+    match = re.search(rf"\#{method}-(\d+)", text)
+    before_match = text[: match.start()]
+    lineno = before_match.count("\n")
+    column = len(before_match.split("\n")[-1]) - int(match.group(1))
+    return lsp.Position(line=lineno, character=column)
+
+
+class ThreadedServer:
+    """
+    Gathers all messages received from server - to handle random-order-messages
+    that are not a response to a request.
+    """
+
+    def __init__(self, process, root_uri):
+        self.process = process
+        self.root_uri = root_uri
+        self.lsp_client = lsp.Client(root_uri=root_uri, trace="verbose")
+        self.lsp_client._recv_catches_and_logs_errors = False
+        self.msgs = []
+
+        self._pout = process.stdout
+        self._pin = process.stdin
+
+        self._read_q = queue.Queue()
+        self._send_q = queue.Queue()
+
+        self.reader_thread = threading.Thread(
+            target=self._read_loop, name="lsp-reader", daemon=True
+        )
+        self.writer_thread = threading.Thread(
+            target=self._send_loop, name="lsp-writer", daemon=True
+        )
+
+        self.reader_thread.start()
+        self.writer_thread.start()
+
+        self.exception = None
+
+    # threaded
+    def _read_loop(self):
+        try:
+            while True:
+                data = self.process.stdout.read(1)
+
+                if data == b"":
+                    break
+
+                self._read_q.put(data)
+        except Exception as ex:
+            self.exception = ex
+        self._send_q.put_nowait(None)  # stop send loop
+
+    # threaded
+    def _send_loop(self):
+        try:
+            while True:
+                chunk = self._send_q.get()
+                if chunk is None:
+                    break
+
+                # print(f"\nsending: {buf}\n")
+                self.process.stdin.write(chunk)
+                self.process.stdin.flush()
+        except Exception as ex:
+            self.exception = ex
+
+    def _queue_data_to_send(self):
+        send_buf = self.lsp_client.send()
+        if send_buf:
+            self._send_q.put(send_buf)
+
+    def _read_data_received(self):
+        while not self._read_q.empty():
+            data = self._read_q.get()
+            events = self.lsp_client.recv(data)
+            for ev in events:
+                self.msgs.append(ev)
+                self._try_default_reply(ev)
+
+    def _try_default_reply(self, msg):
+        if isinstance(msg, lsp.ShowMessageRequest):
+            msg.reply()
+
+    def wait_for_message_of_type(self, type_, timeout=5):
+        end_time = time.time() + timeout
```

Comment:
Maybe a place for [time.monotonic](https://devdocs.io/python~3.9/library/time#time.monotonic)?